File: p1.js
// ## Problem1 
// Pow(x,n) (https://leetcode.com/problems/powx-n/)

// time: O(logN)
// spac: O(logN)

const myPow = function(x, n) {
    // base
    if(n === 0) return 1; 
    // logic
    let temp = myPow(x, parseInt(n/2));
    if(n % 2 === 0) return temp * temp;
    else {
        if(n < 0) return temp * temp * (1/x);
        else return temp * temp * x;
    }
}

File: p1.java
// Time Complexity :O(log(n))
// Space Complexity : O(log(n))
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

class Solution {
    double result;
    public double myPow(double x, int n) {
        if(n == 0) return 1;

//Divide the power in half and get the value recursively
        double temp = myPow(x, n/2);
        //Multiply it by itself to get the result. Check for edge cases ie even and odd power and negative power
        if(n%2 == 1){
            result = temp *temp* x;
        } else if(n%2 == -1){
            result = temp *temp* (1/x);
        }
        else{
            result = temp *temp;
        }
        return result;
    }

    
}